File: backend_python/src/services/mortgage_service.py
"""Mortgage service for database operations."""
import uuid
from typing import Optional
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ..models.mortgage import Mortgage
from ..models.schemas import MortgageCreate, MortgageUpdate


class MortgageService:
    """Service for mortgage-related database operations."""

    async def create_mortgage(self, db: AsyncSession, data: MortgageCreate) -> Mortgage:
        """Create a new mortgage.

        Args:
            db: Database session
            data: Mortgage creation data

        Returns:
            Created mortgage instance
        """
        mortgage = Mortgage(
            id=str(uuid.uuid4()),
            user_id=data.user_id,
            principal=data.principal,
            current_balance=data.current_balance,
            interest_rate=data.interest_rate,
            monthly_payment=data.monthly_payment,
            start_date=data.start_date,
            term_months=data.term_months,
            monthly_income=data.monthly_income,
            monthly_expenses=data.monthly_expenses,
        )
        db.add(mortgage)
        await db.commit()
        await db.refresh(mortgage)
        return mortgage

    async def get_mortgage(self, db: AsyncSession, mortgage_id: str) -> Optional[Mortgage]:
        """Get a mortgage by ID.

        Args:
            db: Database session
            mortgage_id: Mortgage ID

        Returns:
            Mortgage instance or None if not found
        """
        result = await db.execute(select(Mortgage).filter_by(id=mortgage_id))
        return result.scalar_one_or_none()

    async def get_mortgages_by_user(self, db: AsyncSession, user_id: str) -> list[Mortgage]:
        """Get all mortgages for a user.

        Args:
            db: Database session
            user_id: User ID

        Returns:
            List of mortgage instances
        """
        result = await db.execute(select(Mortgage).filter_by(user_id=user_id))
        return list(result.scalars().all())

    async def update_mortgage(
        self, db: AsyncSession, mortgage_id: str, data: MortgageUpdate
    ) -> Optional[Mortgage]:
        """Update a mortgage.

        Args:
            db: Database session
            mortgage_id: Mortgage ID
            data: Update data

        Returns:
            Updated mortgage instance or None if not found
        """
        mortgage = await self.get_mortgage(db, mortgage_id)
        if not mortgage:
            return None

        update_data = data.model_dump(exclude_unset=True)
        for key, value in update_data.items():
            setattr(mortgage, key, value)

        await db.commit()
        await db.refresh(mortgage)
        return mortgage

    async def delete_mortgage(self, db: AsyncSession, mortgage_id: str) -> bool:
        """Delete a mortgage.

        Args:
            db: Database session
            mortgage_id: Mortgage ID

        Returns:
            True if deleted, False if not found
        """
        mortgage = await self.get_mortgage(db, mortgage_id)
        if not mortgage:
            return False

        await db.delete(mortgage)
        await db.commit()
        return True

File: backend/src/services/mortgageService.ts
import { db } from '../models/db';
import { mortgages } from '../models/schema';
import { eq } from 'drizzle-orm';
import { randomUUID } from 'crypto';
import { Mortgage } from '../utils/types';

export class MortgageService {
  async createMortgage(data: {
    userId: string;
    principal: number;
    currentBalance: number;
    interestRate: number;
    monthlyPayment: number;
    startDate: Date;
    termMonths: number;
    monthlyIncome?: number;
    monthlyExpenses?: number;
  }): Promise<Mortgage> {
    const now = new Date();

    console.log('DEBUG: Incoming data:', {
      monthlyIncome: data.monthlyIncome,
      monthlyExpenses: data.monthlyExpenses,
      hasIncome: !!data.monthlyIncome,
      hasExpenses: !!data.monthlyExpenses
    });

    const mortgage = {
      id: randomUUID(),
      userId: data.userId,
      principal: data.principal,
      currentBalance: data.currentBalance,
      interestRate: data.interestRate,
      monthlyPayment: data.monthlyPayment,
      startDate: data.startDate,
      termMonths: data.termMonths,
      monthlyIncome: data.monthlyIncome || null,
      monthlyExpenses: data.monthlyExpenses || null,
      createdAt: now,
      updatedAt: now,
    };

    console.log('DEBUG: Mortgage object to insert:', {
      id: mortgage.id,
      monthlyIncome: mortgage.monthlyIncome,
      monthlyExpenses: mortgage.monthlyExpenses
    });

    await db.insert(mortgages).values(mortgage);

    console.log('DEBUG: Insertion complete, retrieving from DB...');
    const saved = await db.select().from(mortgages).where(eq(mortgages.id, mortgage.id)).limit(1);
    console.log('DEBUG: Saved mortgage from DB:', {
      id: saved[0]?.id,
      monthlyIncome: saved[0]?.monthlyIncome,
      monthlyExpenses: saved[0]?.monthlyExpenses
    });

    return mortgage;
  }

  async getMortgage(id: string): Promise<Mortgage | null> {
    const result = await db.select().from(mortgages).where(eq(mortgages.id, id)).limit(1);
    return result[0] || null;
  }

  async getMortgagesByUser(userId: string): Promise<Mortgage[]> {
    return await db.select().from(mortgages).where(eq(mortgages.userId, userId));
  }

  async updateMortgage(
    id: string,
    data: Partial<{
      currentBalance: number;
      monthlyPayment: number;
      interestRate: number;
    }>
  ): Promise<Mortgage | null> {
    const updatedData = {
      ...data,
      updatedAt: new Date(),
    };

    await db.update(mortgages).set(updatedData).where(eq(mortgages.id, id));
    return this.getMortgage(id);
  }

  async deleteMortgage(id: string): Promise<boolean> {
    await db.delete(mortgages).where(eq(mortgages.id, id));
    return true;
  }
}
